x/chainlink/keeper/keeper.go

```diff
@@ -34,3 +38,95 @@ func NewKeeper(
 func (k Keeper) Logger(ctx sdk.Context) log.Logger {
 	return ctx.Logger().With("module", fmt.Sprintf("x/%s", types.ModuleName))
 }
+
+func (k Keeper) SetFeedData(ctx sdk.Context, feedData *types.MsgFeedData) {
+	// TODO: add more complex feed validation here such as verify against other modules
+
+	// TODO: deserialize the feedData.FeedData if it's an OCR report, assume all the feedData is OCR report for now.
+	// this is simulating the OCR report deserialization lib
+	/****************/
+	observations := make([]*types.Observation, 0, len(feedData.GetFeedData()))
+	for _, b := range feedData.GetFeedData() {
+		o := &types.Observation{Data: []byte(string(b))}
+		observations = append(observations, o)
+	}
+	deserializedOCRReport := types.OCRAbiEncoded{
+		Context:      []byte("testcontext"),
+		Oracles:      feedData.Submitter.Bytes(),
+		Observations: observations,
+		Submitter:    feedData.GetSubmitter(),
+	}
+	/****************/
+	// TODO: verify deserializedOCRReport here
+
+	finalFeedDataInStore := types.OCRFeedDataInStore{
+		FeedData:              feedData,
+		DeserializedOCRReport: &deserializedOCRReport,
+		RoundId:               roundId,
+	}
+	// TODO: add round index here, need to figure out a way to generate round number, below line is a tmp solution
+	roundId += 1
+
+	// use store with gas meter
+	store := ctx.KVStore(k.storeKey)
+
+	f := k.cdc.MustMarshalBinaryBare(&finalFeedDataInStore)
+	store.Set(types.KeyPrefix(types.FeedDataKey), f)
+}
+
+func (k Keeper) GetRoundFeedDataByFilter(ctx sdk.Context, req *types.GetRoundDataRequest) (*types.GetRoundDataResponse, error) {
+	if req == nil {
+		return nil, status.Error(codes.InvalidArgument, "invalid request")
+	}
+
+	var feedRoundData []*types.RoundData
+
+	// use store with gas meter
+	store := ctx.KVStore(k.storeKey)
+
+	pageRes, err := query.Paginate(store, req.Pagination, func(key []byte, value []byte) error {
```

Comment:
just to bring more clarity, what does "page" mean? not sure how the page-indexing system works...

Reply:
this `query.Paginate` is built in SDK, we simply just use it when query round data

Reply:
try to play with it by CLI and you will see